java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java

```diff
@@ -1195,4 +1193,125 @@ public void updateDeviceLangInfo(PremiumUserDevicesRepository.PremiumUserDevice
+		String uniqueToken = UUID.randomUUID().toString();
+		String url = fileName + "_" + uniqueToken;
+		userFile.sharedUrl = url;
+		filesRepository.saveAndFlush(userFile);
+
+		return url;
+	}
+
+	public PremiumUserFilesRepository.UserFile getUserFileBySharedUrl(String token) {
+		return filesRepository.findUserFileBySharedUrl(token);
+	}
+
+	public MapApiController.FileDownloadResult getFile(PremiumUserFilesRepository.UserFile userFile, PremiumUserDevicesRepository.PremiumUserDevice dev) throws IOException {
+		if (userFile == null || dev == null) {
+			return null;
+		}
+		try (InputStream bin = getInputStream(dev, userFile)) {
+			if (bin != null) {
+				InputStream inputStream = new GZIPInputStream(bin);
+				String fileName = URLEncoder.encode(sanitizeEncode(userFile.name), StandardCharsets.UTF_8);
+				return new MapApiController.FileDownloadResult(inputStream, fileName, APPLICATION_OCTET_STREAM_VALUE);
+			}
+		}
+		return null;
+	}
+
+	public GpxFile getFile(PremiumUserFilesRepository.UserFile file) throws IOException {
+		if (file == null || file.data == null) {
+			return null;
+		}
+		try (InputStream inputStream = new GZIPInputStream(new ByteArrayInputStream(file.data));
+		     Source source = new Buffer().readFrom(inputStream)) {
+			GpxFile gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+			if (gpxFile.getError() == null) {
+				return gpxFile;
+			}
+			return null;
+		}
+	}
+
+	public static class FileSharedInfo {
+		public Set<Integer> accessedUsers;
+		public Set<Integer> blackList;
+
+		public FileSharedInfo(Set<Integer> accessedUsers, Set<Integer> blackList) {
+			this.accessedUsers = accessedUsers;
+			this.blackList = blackList;
+		}
+
+		public FileSharedInfo() {
+			this.accessedUsers = new HashSet<>();
+			this.blackList = new HashSet<>();
+		}
+	}
+
+	public boolean saveAccessedUser(PremiumUserDevicesRepository.PremiumUserDevice dev, PremiumUserFilesRepository.UserFile userFile) {
+		final String ACCESSED_USERS = "accessedUsers";
+		FileSharedInfo fileSharedInfo = gson.fromJson(userFile.sharedInfo, FileSharedInfo.class);
+		if (fileSharedInfo == null) {
+			fileSharedInfo = new FileSharedInfo();
+		}
+		if (!fileSharedInfo.blackList.contains(dev.userid)) {
+			if (fileSharedInfo.accessedUsers.add(dev.userid)) {
+				userFile.sharedInfo.add(ACCESSED_USERS, gson.toJsonTree(fileSharedInfo.accessedUsers));
+				filesRepository.saveAndFlush(userFile);
+			}
+			return true;
+		}
+		return false;
+	}
+
+	public List<String> getAccessedUsers(PremiumUserFilesRepository.UserFile userFile) {
+		List<String> accessedUsers = new ArrayList<>();
+		FileSharedInfo fileSharedInfo = gson.fromJson(userFile.sharedInfo, FileSharedInfo.class);
+		if (fileSharedInfo != null && fileSharedInfo.accessedUsers != null) {
+			for (Integer userId : fileSharedInfo.accessedUsers) {
+				PremiumUsersRepository.PremiumUser pu = usersRepository.findById(userId);
+				if (pu != null) {
+					accessedUsers.add(pu.email);
```

Comment:
Emails is the user id, it's ok. Please check google spreadsheets, otherwise we need to introduce nicknames

Reply:
As later we could introduce nicknames it wouldn't matter that method could be easily changed to use nickname instead of email. Now we can introduce column nickname but that method is not needed with new structure